File: Week12_LinkedList/reverse-list-ii.java
/**
 * Definition for singly-linked list. public class ListNode { int val; ListNode
 * next; ListNode() {} ListNode(int val) { this.val = val; } ListNode(int val,
 * ListNode next) { this.val = val; this.next = next; } }
 */

class Solution {
    public ListNode reverseBetween(ListNode head, int left, int right) {
        if (left == right)
            return head;

        ListNode node = new ListNode(0);
        ListNode temp = node;
        node.next = head;

        int i = 1;
        while (i < left) {
            temp = temp.next;
            i++;
        }

        ListNode curr = temp.next;
        while (left < right) {
            ListNode t = curr.next;
            curr.next = t.next;
            t.next = temp.next;
            temp.next = t;
            left++;
        }

        return node.next;
    }
}

File: Week12_LinkedList/reverse-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reverseBetween(self, head: ListNode, left: int, right: int) -> ListNode:
        
        # Nothing to reverse
        if left == right:
            return head
        
        node = temp = ListNode(0)
        node.next = head
        
        i = 1
        while i < left:
            temp = temp.next
            i += 1        
        
        current = temp.next
        while left < right:
            t = current.next
            current.next = t.next
            t.next = temp.next
            temp.next = t
            left += 1
        
        return node.next